dask/dataframe/_hyperloglog.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*
+u"""Implementation of HyperLogLog
+
+This implements the HyperLogLog algorithm for cardinality estimation, found
+in
+
+    Philippe Flajolet, Éric Fusy, Olivier Gandouet and Frédéric Meunier.
+        "HyperLogLog: the analysis of a near-optimal cardinality estimation
+        algorithm". 2007 Conference on Analysis of Algorithms. Nice, France
+        (2007)
+
+"""
+from __future__ import division
+
+import numpy as np
+import pandas as pd
+
+def compute_first_bit(a):
+    # Consider making this less memory-hungry
+    #bits = np.bitwise_and.outer(a, 1<<np.arange(31, -1, -1))
+    bits = np.bitwise_and.outer(a, 1<<np.arange(32))
+    bits = bits.cumsum(axis=1).astype(np.bool)
+    return 33 - bits.sum(axis=1)
+
+def terrible_dataframe_hash(df):
+    """Get an array that is a hash of a numeric dataframe"""
+    hashes = np.full((len(df),), 0x8675309, dtype=np.uint32)
+    for i, (column_label, series) in enumerate(df.iteritems()):
+        this_hash = terrible_series_hash(series)
+
+        num_to_rotate_by = (13 * i) // 32
+
+        rotated_hash = ((this_hash << num_to_rotate_by) |
+                        (this_hash >> num_to_rotate_by))
+
+        hashes ^= rotated_hash
+
+    return hashes
+
+def terrible_series_hash(s):
+    # Get the column as an uint32 array
+    if s.dtype == object:
+        # TODO: strings
+        raise TypeError("object dtype not supported")
+    if s.dtype.itemsize in (1, 2):
+        s = s.astype(np.int32).view(np.uint32)
+
+    if s.dtype.itemsize in (4, 8, 16):
+        # Get an array of uint32s from floats, complex, ints, etc.
+        width = s.dtype.itemsize // 4 # length of axis 1
+        data = s.values.view(np.uint32).reshape((len(s), width))
+        data *= 0xDEECE66D # Some prime
+
+        # Different offset per column, make 1+2j not same as 2+1j
+        data += np.arange(0xB, 0xB+width, dtype=np.uint32).reshape((1, width))
+
+        # reduce to just one uint32 per entry
+        ints = np.bitwise_xor.reduce(data, axis=1)
+
+        return ints
+
+    raise TypeError("Don't know what to do with {0!r} ({0.dtype})".format(s))
+
+def terrible_hash(obj):
+    return terrible_dataframe_hash(pd.DataFrame(obj))
+
+def state_from_frame(obj, p):
+    return state_from_hashes(terrible_hash(obj), p)
```

Comment:
It's mostly a matter of the order in which I wrote it. (e.g. I defined terrible_hash right at the end when I realized I could have one method defined in _Frame for both Series/DataFrame, but I didn't want to put the `pd.DataFrame(...)` right inline as I was worried the purpose it had would be too subtle there; a comment would suffice too for such subtlety.) I figured the hashing level would be ripped out entirely and replaced with something lacking the adjective "terrible", so I didn't bother cleaning up the layering yet. I didn't necessarily know whether the hashing strategy would ultimately come from the user or be hard-coded or vary on different stuff, etc.